design/libraries/nodes-and-kinds.md

```diff
@@ -0,0 +1,152 @@
+a cleverer way for their compiler and type system to support this.)
+
+### different implementors of Node
+
+Though the methods on the `Node` interface are defined as those necessary for
+examining data of the [Data Model Kinds](../../data-model-layer/data-model.md#kinds),
+**`Node` is not only implemented by the Data Model**:
+
+- Yes, `Node` is implemented by types that just hold basic Data Model info;
+- `Node` is also implemented by [Advanced Data Layouts](../../schemas/advanced-layouts.md) --
+	- consider a HAMT that spans many separately-seralized chunks of data; it should still be usable as if it's a regular map.
+- `Node` is also implemented by [Schema-typed Nodes](../../schemas/) --
+	- Both if implemented by a single implementation that evalates rules at runtime (so, finite count of implementing types and known at core library compile time)...
+	- or if handled by codegen/macros (unknown count / open set of implementors of `Node`; not known at core library compile time; may be created in other packages that import the core, rather than core importing them!).
+
+Even further, some libraries may choose to make even more various
+implementations of `Node` for optimizing performance of specific tasks:
+for example, a `Node` which implements basic Data Model "map" semantics,
+but using some internal algorithm for memory layout which is known to be
+efficient for certain workloads;
+or for another example, a `Node` which is particularly efficient for handling
+data of one particular serialization codec, and keeps a lazy-loading skip-tree
+over the serialized bytes.
+Clearly, neither of these should be the default implementation a library uses,
+but clearly, both of them should be able to be used transparently,
+
+With all seven (?! indeed, *seven*) of these different stories,
+we can consider it conclusive that the `Node` interface should be ready
+to support many, many diverse implementors.
+
+### a default implementation of Node
+
+As an IPLD library author, you may be tempted to make a single, "default"
+implementation of `Node`.
+
+Feel free to do so; but be cautious of giving it special privileges.
+Try implementing it in a separate package from your core interfaces: this will
+be a good exercise to make sure other implementations can later do the same.
+(Since in the order of things you'll do when implementing a new IPLD library,
+creating this basic default node implementation is likely quite early,
+going about it in such a way that it forces design choices you'll need later
+anyway will save you from potentially discovering the need for some costly
+refactors later!)
+
+
+Higher level functions
+----------------------
+
+Almost all features should be implemented to take `Node` arguments,
+and return `Node` values.
+
+Traversals and walks can be implemented in this way.
+
+Selectors can be implemented in this way.
+
+Transformations can be implemented in this way.
+
+Codecs themselves can be implemented this way.
```

Comment:
Can you expand on this one? I can't even figure out what it would mean for a codec to take a `Node` and return a `Node`.

Reply:
Added some psuedocode to explore this.  I'm worried it might be distracting line noise if not read tolernately, but I guess it's probably better than nothing.